x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sample/SampleIterator.java

```diff
@@ -174,6 +183,14 @@ private void queryForCompositeAggPage(ActionListener<Payload> listener, final Sa
         }));
     }
 
+    private void addShardFailures(SearchResponse r) {
+        if (r.getShardFailures() != null) {
+            for (ShardSearchFailure shardFailure : r.getShardFailures()) {
+                shardFailures.put(shardFailure.toString(), shardFailure); // TODO find a better way to deduplicate
```

Comment:
Or limit the number of shard failures. How does ES deal with this scenario?

Reply:
ES returns all the failures, and I tend to agree with this approach (returning complete information).
ShardSearchFailure doesn't have an equals/hashcode implementation, and since it's widely used, I'm not sure it's something we want to introduce now.